src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1626,54 +1655,64 @@ PlanRouterQuery(Query *originalQuery, RelationRestrictionContext *restrictionCon
 	 * still run the query but the result will be empty. We create a dummy shard
 	 * placement for the first active worker.
 	 */
-	if (shardsPresent)
+	if (!localMultiShardModifyQuery)
 	{
-		workerList = WorkersContainingAllShards(prunedRelationShardList);
-	}
-	else if (replacePrunedQueryWithDummy)
-	{
-		List *workerNodeList = ActiveReadableNodeList();
-		if (workerNodeList != NIL)
+		if (shardsPresent)
 		{
-			WorkerNode *workerNode = (WorkerNode *) linitial(workerNodeList);
-			ShardPlacement *dummyPlacement =
-				(ShardPlacement *) CitusMakeNode(ShardPlacement);
-			dummyPlacement->nodeName = workerNode->workerName;
-			dummyPlacement->nodePort = workerNode->workerPort;
-			dummyPlacement->groupId = workerNode->groupId;
+			workerList = WorkersContainingSelectShards(prunedRelationShardList);
+		}
+		else if (replacePrunedQueryWithDummy)
+		{
+			List *workerNodeList = ActiveReadableNodeList();
+			if (workerNodeList != NIL)
+			{
+				WorkerNode *workerNode = (WorkerNode *) linitial(workerNodeList);
+				ShardPlacement *dummyPlacement =
+					(ShardPlacement *) CitusMakeNode(ShardPlacement);
+				dummyPlacement->nodeName = workerNode->workerName;
+				dummyPlacement->nodePort = workerNode->workerPort;
+				dummyPlacement->groupId = workerNode->groupId;
+
+				workerList = lappend(workerList, dummyPlacement);
+			}
+		}
+		else
+		{
+			/*
+			 * For INSERT ... SELECT, this query could be still a valid for some other target
+			 * shard intervals. Thus, we should return empty list if there aren't any matching
+			 * workers, so that the caller can decide what to do with this task.
+			 */
+			return NULL;
+		}
 
-			workerList = lappend(workerList, dummyPlacement);
+		if (workerList == NIL)
+		{
+			ereport(DEBUG2, (errmsg("Found no worker with all shard placements")));
+
+			planningError = DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+										  "found no worker with all shard placements",
+										  NULL, NULL);
+			return planningError;
 		}
 	}
 	else
 	{
-		/*
-		 * For INSERT ... SELECT, this query could be still a valid for some other target
-		 * shard intervals. Thus, we should return empty list if there aren't any matching
-		 * workers, so that the caller can decide what to do with this task.
-		 */
-		return NULL;
+		workerList = WorkersContainingModifyShards(prunedRelationShardList);
```

Comment:
What's the point of this code? We don't use the result as far as I can tell (`placementList`) isn't used for multi-shard router queries in `RouterJob`. Is this something we can omit?